rust/plugin-lib/src/base_cache.rs

```diff
@@ -27,49 +30,270 @@ const CHUNK_SIZE: usize = 1024 * 1024;
-                // At least the first codepoint at start is in the chunk.
-                if end < chunk_end || chunk_end == self.buf_size {
-                    return Ok(&self.contents[start - chunk_start ..]);
-                }
-                let new_chunk = source.get_data(chunk_end, CHUNK_SIZE, self.rev)?;
-                if start == chunk_start {
-                    self.contents.push_str(&new_chunk);
-                } else {
-                    self.offset = start;
-                    self.contents = [&self.contents[start - chunk_start ..],
-                                     &new_chunk].concat();
-                }
-            } else {
-                // TODO: if chunk_start < start + CHUNK_SIZE, could fetch smaller
-                // chunk and concat; probably not a major savings in practice.
-                self.contents = source.get_data(start, CHUNK_SIZE, self.rev)?;
-                self.offset = start;
+            if let Some(end_off) = self.cached_offset_of_line(line_num + 1) {
+                return Ok(&self.contents[start_off..end_off])
+            }
+            // if we have a chunk and we're fetching more, discard unnecessary
+            // portion of our chunk.
+            if start_off != 0 {
+                self.clear_up_to(start_off);
+                start_off = 0;
+            }
+
+            let chunk_end = self.offset + self.contents.len();
+            let resp = source.get_data(chunk_end, TextUnit::Utf8,
+                                       CHUNK_SIZE, self.rev)?;
+            self.append_chunk(resp);
+        }
+    }
+
+    /// Returns the offset of the line at `line_num`, zero-indexed, fetching
+    /// data from `source` if needed.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if `line_num` is greater than the total number of lines
+    /// in the document, or if there is a problem communicating with `source`.
+    pub fn offset_of_line<DS>(&mut self, source: &DS, line_num: usize) -> Result<usize, Error>
+        where DS: DataSource
+    {
+        if line_num > self.num_lines { return Err(Error::BadRequest) }
+        match self.cached_offset_of_line(line_num) {
+            Some(offset) => Ok(offset + self.offset),
+            None => {
+                let resp = source.get_data(line_num, TextUnit::Line, CHUNK_SIZE, self.rev)?;
+                self.reset_chunk(resp);
+                self.offset_of_line(source, line_num)
             }
         }
     }
 
+    /// Returns the offset of the provided `line_num` in `self.contents` if
+    /// it is present in the chunk.
+    fn cached_offset_of_line(&self, line_num: usize) -> Option<usize> {
+        if line_num < self.first_line { return None }
+
+        let rel_line_num = line_num - self.first_line;
+
+        if rel_line_num == 0 && self.first_line_offset == 0 {
+            return Some(0)
+        }
+        if rel_line_num <= self.line_offsets.len() {
```

Comment:
So we've got a problem if rel_line_num is 0 and first_line_offset is not 0; this code will panic because the `- 1` will underflow. We actually know the answer; it's `-self.first_line_offset`, but that won't fit in a `usize` as it's negative.

Reply:
good catch. fixed this up and added a test case.